nixos/modules/services/audio/jack.nix

```diff
@@ -0,0 +1,275 @@
+              type asym
+              playback.pcm "asoftvol"
+              capture.pcm "loophw01"
+            }
+            pcm.!default {
+              type plug
+              slave.pcm aduplex
+            }
+          '';
+        };
+
+        session = mkOption {
+          type = types.string;
+          default = ''
+            alsa_in -j cloop -dcloop &
+            alsa_out -j ploop -dploop &
+            while [ "$(jack_lsp cloop)" == "" ] || [ "$(jack_lsp ploop)" == "" ]; do sleep 1; done
+            jack_connect cloop:capture_1 system:playback_1
+            jack_connect cloop:capture_2 system:playback_2
+            jack_connect system:capture_1 ploop:playback_1
+            jack_connect system:capture_2 ploop:playback_2
+          '';
+          description = ''
+            Additional commands to run to setup loopback device.
+          '';
+        };
+      };
+
+    };
+
+  };
+
+  config = mkMerge [
+
+    (mkIf pcmPlugin {
+      sound.extraConfig = ''
+        pcm_type.jack {
+          libs.native = ${pkgs.alsaPlugins}/lib/alsa-lib/libasound_module_pcm_jack.so ;
+          ${lib.optionalString enable32BitAlsaPlugins
+          "libs.32Bit = ${pkgs.pkgsi686Linux.alsaPlugins}/lib/alsa-lib/libasound_module_pcm_jack.so ;"}
+        }
+        pcm.!default {
+          @func getenv
+          vars [ PCM ]
+          default "plug:jack"
+        }
+      '';
+    })
+
+    (mkIf loopback {
+      boot.kernelModules = [ "snd-aloop" ];
+      boot.kernelParams = [ "snd-aloop.index=${toString cfg.loopback.index}" ];
+      sound.extraConfig = cfg.loopback.config;
+    })
+
+    (mkIf cfg.jackd.enable {
+     assertions = [
+        {
+          assertion = !(cfg.alsa.enable && cfg.loopback.enable);
+          message = "Both alsa and loopback options shouldn't be used at the same time.";
+        }
+      ];
+
+      users.users.jackaudio = {
+        group = "jackaudio";
+        extraGroups = [ "audio" ];
+        description = "JACK Audio system service user";
+      };
+      # http://jackaudio.org/faq/linux_rt_config.html
+      security.pam.loginLimits = [
+        { domain = "@jackaudio"; type = "-"; item = "rtprio"; value = "99"; }
+        { domain = "@jackaudio"; type = "-"; item = "memlock"; value = "unlimited"; }
+      ];
+      users.groups.jackaudio = {};
+
+      environment = {
+        systemPackages = [ cfg.jackd.package ];
+        etc."alsa/conf.d/50-jack.conf".source = "${pkgs.alsaPlugins}/etc/alsa/conf.d/50-jack.conf";
+        variables.JACK_PROMISCUOUS_SERVER = "jackaudio";
+      };
+
+      services.udev.extraRules = ''
+        ACTION=="add", SUBSYSTEM=="sound", ATTRS{id}!="Loopback", TAG+="systemd", ENV{SYSTEMD_WANTS}="jack.service"
+      '';
+
+      systemd.services.jack = {
+        description = "JACK Audio Connection Kit";
+        serviceConfig = {
+          User = "jackaudio";
+          ExecStart = "${cfg.jackd.package}/bin/jackd ${concatStringsSep " " cfg.jackd.extraOptions}";
```

Comment:
Use `lib.escapeShellArgs` instead of `concatStringsSep " "`

Reply:
@Infinisil Hmm, this way it generates `ExecStart=/nix/store/inqg0qsbd5z0i8l4515n3h800dm6di82-jack2-1.9.12/bin/jackd '-dalsa' '--device hw:1'` instead of `ExecStart=/nix/store/inqg0qsbd5z0i8l4515n3h800dm6di82-jack2-1.9.12/bin/jackd -dalsa --device hw:1` and so jackd fails to start

Reply:
Hmm I see, you'd have to change the declaration to `[ "-dalsa" "--device" "hw:1" ]`. Often CLI's support leaving out the space even, so maybe you can do `[ "-dalsa" "--devicehw:1" ]`.